### ISSUE TITLE
Can we send the cookie with request in Nuke?

### ISSUE THREAD

**Issue by poorni008** [kean/Nuke#368]
I want to know whether the cookie is supported in Nuke.I want to send a cookie along with URL,is that possible in Nuke?.

**Comment by kean** [comment]
Hi @poorni008. If the cookies are added to `HTTPCookieStorage.shared` for the needed domain, they should automatically be attached to the requests. You can add cookies manually to the storage.

**Comment by poorni008** [comment]
Hi Kean,Thanks for your response..i have tried with this code..func request(with url: URL) -> URLRequest {
      var request = URLRequest(url: url)

        guard let cookies = HTTPCookieStorage.shared.cookies(for: url) else {
            return request
        }

        request.allHTTPHeaderFields = HTTPCookie.requestHeaderFields(with: cookies)
        return request
    } .I want to know how to call the DataLoader methods..Is there any sample code.I want to fetch the image from my CMS.

**Comment by kean** [comment]
If the cookies are in the cookie storage and are registered for the correct domains, it should just work, they should get attached automatically. If you want to set HTTP headers fields manually, please refer to https://github.com/kean/Nuke/blob/master/Documentation/Guides/faq.md.

**Comment by poorni008** [comment]
Thanks for your response...i want to send the session id in cookie..i will try with code of data loader.

**Comment by poorni008** [comment]
Hi kean,I need one more help on creating a session using framework and get the image as data.I want the image as data in response.It would be helpful, if you share example code.

**Comment by kean** [comment]
Could you please clarify what is the use case? There are a few different ways to get access to image data, it depends when and what you need it for.

**Comment by poorni008** [comment]
I want to send the request using URLSession and get the response back from server as data.The data I will convert into image and save it in document folder...next time I don’t want to download the image again I will fetch from document this is my usecase.This I need to achieve using nuke.one more doubt if I didn’t mention any configuration in  image pipeline does it take default configuration.

**Comment by kean** [comment]
There is a `DataCaching` protocol which you can implement to inject your custom cache in the pipeline. There is also an existing [`DataCache`](https://github.com/kean/Nuke#aggressive-lru-disk-cache) class that implements this protocol and that you can use to save images persistently. This way you are going to be taking advantage of the entire Nuke infrastructure, but you are still going to achieve your goal of saving images in documents folder.

> one more doubt if I didn’t mention any configuration in image pipeline does it take default configuration.

Sorry, I'm not sure I follow. If you are talking about the `ImagePipeline` configuration, then yes, there is a default. Here an example of how to create a pipeline with the default configuration and override the parts that you need:

```swift
ImagePipeline {
    $0.dataCache = try? DataCache(name: "com.myapp.datacache")

    // Also consider disabling the native HTTP cache, see `DataLoader`.
}
```

**Comment by poorni008** [comment]
Hi kean,Thanks for the response..i have commented this line of code 
/* let  pipeline = ImagePipeline {
            let urlSessionConf = URLSessionConfiguration.default
            urlSessionConf.urlCache = nil // disable disk cache
            $0.dataLoader = DataLoader(configuration: urlSessionConf)}*/
and just called 
ImagePipeline.shared.loadData(with: imagerquest, completion: { [weak self] result in
            guard let _ = self,let data = try? result.get().data else
            {
                return
            }
            print("task completed")
            print(result);
            print(data)
        })
still getting the same output..i just want to want to know if we didn't configure anything in ImagePipeline...it takes default data loader,datacache?.I want to maintain the session in my app..that is possible via Dataloader class?

**Comment by kean** [comment]
The pipeline configuration looks good, but you should set `dataCache`. I don't see where you are setting it.

The call to `loadData()` also looks fine. Having said that, I would suggest using `loadImage()` if possible.

**Comment by poorni008** [comment]
`@objc public func setURLToImage(urlstring:String,backgroundimageview:UIImageView)
    {
        let pipeline = ImagePipeline
        {
            let urlSessionConf = URLSessionConfiguration.default
            urlSessionConf.urlCache = nil // disable disk cache
            $0.dataLoader = DataLoader(configuration: urlSessionConf)
            $0.dataCache = try? DataCache(name: "com.test.datacache")
            
        }
        ImagePipeline.shared = pipeline
        
        let cache = try! DataCache(name: "com.test.datacache")
        let url1 = NSURL.init(string:"https://i.picsum.photos/id/866/200/200.jpg")
        let request1 = request(with: url1! as URL)
        let imagerquest = ImageRequest.init(urlRequest: request1)
        let key = ImagePipeline.shared.cacheKey(for: imagerquest, item:.originalImageData)
        if let imagedata = cache.cachedData(for: key)
        {
            let idecode = ImageDecoder.init()
            let imagecontainer = idecode.decode(imagedata)
            let cacheimage =  imagecontainer?.image
            backgroundimageview.nuke_display(image: cacheimage)
            
        }
        else
        {
            ImagePipeline.shared.loadImage(with: imagerquest) { [weak self] result in
                print("result",result)
                
                guard let _ = self, let dataimage = try? result.get().container.image else {
                    return
                }
                
                cache.storeData(dataimage.pngData()!, for: key)
                backgroundimageview.nuke_display(image: dataimage)
                
            }
        }
    }
    `

**Comment by kean** [comment]
hey,

```swift
    let key = ImagePipeline.shared.cacheKey(for: imagerquest, item:.originalImageData)
    if let imagedata = cache.cachedData(for: key)
    {
        let idecode = ImageDecoder.init()
        let imagecontainer = idecode.decode(imagedata)
        let cacheimage =  imagecontainer?.image
        backgroundimageview.nuke_display(image: cacheimage)
    }
```

You don't need to write this code manually to use custom disk caching. You only need a single line to enable it, [documented here](https://github.com/kean/Nuke#aggressive-lru-disk-cache). I would also suggest checking the respective demo section.

**Comment by poorni008** [comment]
Hi,
As per my understanding you are saying data cache configuration in Pipline is enough am I right?.I thought if the image is not there in cache load it using imagepipline and if image is there in cache fetch from it and display it to user,in that way  I coded.

**Comment by poorni008** [comment]
how to integrate our own networking layer with nuke.Sample code if any would be useful

**Comment by kean** [comment]
Hi, I appreciate you being interested in integrating this framework, however, please understand that this is not a support forum. Most of the questions you asked so far are covered nicely in the documentation, including your last question regarding networking layer – [see this guide](https://github.com/kean/Nuke/blob/master/Documentation/Guides/third-party-libraries.md).

Please, before you open issues, take your time and see if the answer is already in the documentation. If it's not in the documentation, I would appreciate PRs that improve it. If  something isn't working according to the documentation, feel free to open issues, but please make sure to clearly describe what the problem is: what you did, and what the expected and actual results were.

I'm going to close this issue. I wish I had infinite resources to personally help everyone using this framework, but is it not feasible. I'm going to consider adding a technical support option with a reasonable hourly rate in the future. I appreciate your understanding.
